### ISSUE TITLE
[4.0] Installation proposing MySQL (PDO) as first choice prevents installing Joomla. Choosing this db type does not create the db.

### ISSUE THREAD

**Issue by infograf768** [joomla/joomla-cms#19779]
Very much used to mysqli being proposed as first choice for the database type (as most Joomla users use that db type), I did not care about the field.

In fact the default field value is now MySQL (PDO).
<img width="732" alt="screen shot 2018-02-25 at 10 45 14" src="https://user-images.githubusercontent.com/869724/36640116-8e6cb51e-1a19-11e8-85f8-256a8754a9c5.png">

It totally broke the installation here.

To get back mysqli as first choice, I had to modify the xml to set mysqli as first choice.
```
		<field
				name="db_type"
				type="databaseconnection"
				id="db_type"
				class="custom-select form-control"
				label="INSTL_DATABASE_TYPE_DESC"
				supported="mysql,mysqli,pgsql,postgresql"
				required="true"
				default="mysqli" /// Changed here
				filter="string"
		/>
```
No idea if this is the real solution to that issue as I am not sure about the supposed use of MySQL (PDO) although I read http://php.net/manual/en/ref.pdo-mysql.php

**Comment by infograf768** [comment]
Error log shows this
```
[25-Feb-2018 10:59:53 Europe/Berlin] PHP Warning:  Uncaught PDOException: SQLSTATE[HY000] [1049] Unknown database 'mysqlpdotest' in /Applications/MAMP/htdocs/newfolder/joomla40/libraries/vendor/joomla/database/src/Pdo/PdoDriver.php:312
Stack trace:
#0 /Applications/MAMP/htdocs/newfolder/joomla40/libraries/vendor/joomla/database/src/Pdo/PdoDriver.php(312): PDO->__construct('mysql:host=loca...', 'root', 'root', Array)
#1 /Applications/MAMP/htdocs/newfolder/joomla40/libraries/vendor/joomla/database/src/Mysql/MysqlDriver.php(124): Joomla\Database\Pdo\PdoDriver->connect()
#2 /Applications/MAMP/htdocs/newfolder/joomla40/libraries/vendor/joomla/session/src/Handler/DatabaseHandler.php(222): Joomla\Database\Mysql\MysqlDriver->connect()
#3 [internal function]: Joomla\Session\Handler\DatabaseHandler->open('/Applications/M...', 'f16b9fffa4b4d08...')
#4 /Applications/MAMP/htdocs/newfolder/joomla40/libraries/vendor/joomla/session/src/Storage/NativeStorage.php(454): session_start()
#5 /Applications/MAMP/htdocs/newfolder/joomla40/libraries/src/Session/Storage/JoomlaStorage.p in /Applications/MAMP/htdocs/newfolder/joomla40/libraries/vendor/joomla/database/src/Pdo/PdoDriver.php on line 317
[25-Feb-2018 10:59:53 Europe/Berlin] PHP Fatal error:  session_start(): Failed to initialize storage module: user (path: /Applications/MAMP/tmp/php) in /Applications/MAMP/htdocs/newfolder/joomla40/libraries/vendor/joomla/session/src/Storage/NativeStorage.php on line 454

**Comment by infograf768** [comment]
Looks like it is related to https://github.com/joomla/joomla-cms/pull/19646

**Comment by ghost** [comment]
closed as having Pull Request #19780

**Comment by infograf768** [comment]
i think this should be reopened as the default db proposed is also broken

**Comment by brianteeman** [comment]
I changed the default in #19780

**Comment by mbabker** [comment]
@infograf768 Please do not randomly link unrelated issues, none of the items you linked has ANYTHING to do with what you are describing in this item.

**Comment by mbabker** [comment]
> default db proposed is also broken

No, it's not.  The PDO driver works just fine and I have it in use on production sites.  What is broken is you have something triggering a session in the scope of the install application trying to use the database handler which is impossible.

**Comment by infograf768** [comment]
@mbabker 
what about the error i got. look at the first line above. i can read mysqlpdotest

**Comment by infograf768** [comment]
for me, again a cultural issue?, this shows as broken. i have no idea what is

**Comment by mbabker** [comment]
The problem is not that it is trying to read the database.  The problem is the session handler is trying to connect to the database.  In the install application.  When you do not have a configuration saved.  And can't reliably connect to your database.  That is unrelated to the order of options.

The installation application should always use the filesystem session handler.

As for this issue "PHP Fatal error:  session_start(): Failed to initialize storage module: user (path: /Applications/MAMP/tmp/php)" this implies that the MAMP temp directory doesn't have suitable read/write permissions.  https://github.com/joomla/joomla-cms/commit/efc468ab2ef9301498e99ee932ab051877453cdd might fix that by making the configuration try to use the PHP runtime configuration instead of completely ignoring it, no promises on that one though.

As for what PDO is, it's basically another database connector built in C as an extension of PHP itself.  Similar to how there are different extensions for legacy MySQL, MySQLi, and PostgreSQL.  PDO tries to act as an abstraction layer (similar to what our database API does) and allows connections to different database engines.  In 4.0 we explicitly support both MySQL and PostgreSQL connections either through the native PHP extensions (`ext/mysqli` or `ext/pgsql`) or the PDO abstraction layer.  The only thing that really changes at our level is what internal PHP code is executed.

**Comment by infograf768** [comment]
@mbabker 
I know it is not related to the order of Options as mysqli works perfectly.

Some more infos;

1. The session is perfectly saved in Applications/MAMP/tmp/php (here my last test is at 8:07). the folder has correct permissions.
Error: 
`[26-Feb-2018 08:07:57 Europe/Berlin] PHP Fatal error:  session_start(): Failed to initialize storage module: user (path: /Applications/MAMP/tmp/php) in /Applications/MAMP/htdocs/newfolder/joomla40/libraries/vendor/joomla/session/src/Storage/NativeStorage.php on line 454`

![screen shot 2018-02-26 at 08 15 25](https://user-images.githubusercontent.com/869724/36657526-456ff010-1acd-11e8-834a-b0dbd9755482.png)

2 . A configuration.php is saved. It contains peculiarly for this 8:07 test:
```
public $dbtype = 'mysql';
	public $host = 'localhost';
	public $user = 'root';
	public $password = 'root';
	public $db = 'dbtest';
	public $dbprefix = 'jos_';
```

But the db is NOT created.
<img width="722" alt="screen shot 2018-02-26 at 08 20 31" src="https://user-images.githubusercontent.com/869724/36657737-fecbfd6a-1acd-11e8-9d3b-c9d0ba20d8b1.png">

**Comment by infograf768** [comment]
I modified the title of this issue to fit.

**Comment by infograf768** [comment]
@mbabker 
I found the issue.
One has to first manually create the db before installing.
I.e. it is not created on the fly as msqli does.

Is that expected or is it a bug?

**Comment by infograf768** [comment]
@laoneo  I am told to ping you.

**Comment by mbabker** [comment]
Change the type at https://github.com/joomla/joomla-cms/blob/4.0-dev/installation/src/Model/DatabaseModel.php#L343 to "mysql" instead of "pdomysql".

**Comment by infograf768** [comment]
> Change the type at https://github.com/joomla/joomla-cms/blob/4.0-dev/installation/src/Model/DatabaseModel.php#L343 to "mysql" instead of "pdomysql".

This works, but should it no be also modified elsewhere in core?

https://github.com/joomla/joomla-cms/blob/4.0-dev/installation/src/Model/SetupModel.php#L464
(That one looks a bit redundant with the same method in DatabaseModel)

Also
https://github.com/joomla/joomla-cms/blob/4.0-dev/installation/src/Form/Field/Installation/SampleField.php#L44
https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/Installer/Installer.php#L954
https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/Installer/Installer.php#L1121

What about the code in libraries/src/Service/Provider/Database.php?
And also in test pdomysql.php?

**Comment by mbabker** [comment]
You know, at one point I made all the required changes for that move, so I really don't know why all these problems are back.

**Comment by infograf768** [comment]
I'll leave these modifications to the powers that be.
For example, it may be that the code in libraries/src/Service/Provider/Database.php is there for B/C.
Over my abilities.

**Comment by mbabker** [comment]
Read the inline commentary in the database provider, it is a B/C layer 😉

**Comment by mbabker** [comment]
Well, at least the changes I made to the installer in https://github.com/joomla/joomla-cms/pull/16402 got lost.  So once again someone is going to have to scan the full code base and adapt for the changed code structure.

**Comment by laoneo** [comment]
Can the problem be the move to namespaces in the installation app? Did I screw something up?

**Comment by infograf768** [comment]
my tests show that using mysql instead of pdomysql works fine. The xml does not use pdomysql.

**Comment by brianteeman** [comment]
The xml should not use pdoxml as that would be invalid

**Comment by mbabker** [comment]
> Did I screw something up?

When the install app was converted to its current state some of the changes I made related to moving to the Framework's database package were lost (all the code references to "pdomysql" should have been replaced as seen in #16402 but looking at the current code they're back).  So like I said, we just need to do another scan on the code base and find these remaining references and fix them.

**Comment by mbabker** [comment]
No guarantees but https://github.com/joomla/joomla-cms/pull/19989 might fix this.

**Comment by ghost** [comment]
closed as having Pull Request #19989
